### PR TITLE
Creates Advanced JavaScript curriculum

### PR DESCRIPTION
**Pull request by raorao**
Students are asked to create a to do list application that consumes an already written and deployed API. The application uses React.js, using a pre-written skeleton that includes local copies of all dependencies. HTML and CSS are also pre-written. This is a tutorial meant for students with some exposure to JavaScript.  See curriculum for more details. 

Requirements:
* requires an operational listalous.herokuapp.com (as an API). see on github at https://github.com/raorao/headless_todos.
* asks students to download and unzip resources from https://github.com/raorao/AdvancedTodoList .

### CODE REVIEW

**Comment by tjgrathwell** [comment]
Looks interesting. "Reactifying" all that HTML really makes it look a lot longer and scarier to me :ghost: 

Bugs:
* The `loading_items` page instructs the student to add `ListStore.loadItems()` to the bottom of their index.html file, but in the current base project it's already there.
* `loading_items` has a "Step 1, Step 1.1, Step 1.2"; I assume this should just be "Step 1" "Step 2" "Step 3" instead.
* `adding_an_item` has two "Steps:" headers
* There's a `playing_with_javascript` page in the advanced JS site but it's not linked. Is that intentional?

Thoughts:
* I noticed that in the intermediate JS curriculum jQuery is loaded off a CDN, but here all the deps are vendored. Any reason for that? It might be easier to pull the sample app code into the docs repo if there's less dependencies.
* All the JSX content is one big `<script>` tag in index.html. Is it possible/desirable to move this to a separate file? Maybe different files for each component?
* Should we find a way to highlight JSX better? It looks like CodeRay doesn't quite understand it.
* The React version clears out the text in the create form after submit, while the jQuery one does not (when doing the jQuery version I thought clearing out the form might be an extra credit activity.). Should they both clear out the form? Is `this.refs.description.getDOMNode().value = ''` really the Reactiest way to do that? I don't know much about React but that seems goofy.
* I found it hard to follow on the `adding_an_item` page that I was being asked to add an `onClick` attribute, specifically, because it just dumps a bunch of updated JSX in my face.
* Why is `findItemById` just floating out there rather than being a member of `ListSource`?

**Comment by raorao** [comment]
Thanks for reviewing, @tjgrathwell ! Responding to your comments:

* whoops on the ListStore.addItem() snafu. That's fixed now. good catch. 
* fixed unintended step nesting in a3be960
* removes secondary steps header in 8153c16
* playing_with_javascript was a late removal. I've removed the file in 01b1780

* I'm loading the react dependencies locally as well, since IIRC they are required to run locally for an out-of-the-box setup. I didn't seem write to have some dependencies local, and others be CDNs. That said, I've got no strong opinions on this. 
* Unfortunately, Facebook's in-browser JSX transformer does not work with external script files. A bummer, I know. 
* I agree, the lessons should be consistent. The jQuery one should probably also clear out the form. 
* Haha I know, that whole handler looks a little awkward. The refs.refName.getDOMNode() pattern is not particularly visually appealing, but it's pretty widely used across Facebook's official docs. Using built in JS functions to manipulate DOM elements is also common, including the ```value = ''``` technique. http://facebook.github.io/react/docs/tutorial.html#adding-new-comments . FWIW, my company's code (which is very React-y) uses the same technique. 
* Totally fair on the JSX pile. I've changed the structure -- now the reader is asked to just alter that line code, followed by a section of "here's what your code should look like." changed in 5eb52db
* I wanted to prevent readers from using ```this``` from within their code, and definitely protect them from binding context, which would have been required of findItemById was defined on ListStore object. IRL, I would probably write this using a closure, but that's another concept I wanted to protect students from. The compromise solution was these awkward global functions. My thinking is that readers could choose to closure-ize store.js in the future and preserve functionality. Happy to hear any other ideas, though.

**Comment by raorao** [comment]
@tjgrathwell would it be possible to merge this in provisionally? I'd love to have this to show students/volunteers.

**Comment by tjgrathwell** [comment]
Sure thing, here it is right now on this website: http://docs.railsbridge.org/javascript-to-do-list-with-react

Another question, do you imagine that the "Dev tools" and "Creating a list" pages will always remain the same between the two curricula? If so, we could probably just have one reference the other via `insert` like this: https://github.com/railsbridge/docs/blob/master/sites/en/job-board/job-board.step#L28
